src/racer/ast_types.rs

```diff
@@ -449,11 +470,13 @@ impl fmt::Display for Path {
 pub struct PathSegment {
     pub name: String,
     pub generics: Vec<Ty>,
+    /// If this path segment is a closure, it's return type
+    pub output: Option<Ty>,
```

Comment:
I think 
```
generics: Option<GenericsKind>,
...
enum GenericsKind {
   Angle(Vec<Ty>),
   Paren{in: Vec<Ty>, out: Option<Ty> },
}
```
is better.
But, since it needs large refactoring, you don't have to fix it in this PR.
I'll fix it later.